### ISSUE TITLE
ENH: return IPython rich display objects

### ISSUE THREAD

**Issue by westurner** [looooo/jupyter_freecad_kernel#3]
Users can code and render FreeCAD python in a Jupyter notebook, in order to:
- share designs
- teach design and the FreeCAD API
- 

IPython and Jupyter Notebook support a rich display object protocol that checks for ``._repr_[format]_()`` methods before going to plaintext `.__repr__()`.

https://ipython.readthedocs.io/en/stable/config/integrating.html#rich-display

https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.DisplayObject

https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.HTML
- HTML
  - Heading (and caption) with scalable font size
    - Markdown
      - https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.Markdown
  - [ ] Render and display multiple default perspectives of e.g. parametric designs w/ cadquery
    - In order to embed the image in the HTML, is it necessary to inline base64-encode the png?
    - [ ] `` _repr_mimebundle_()``
      - https://ipython.readthedocs.io/en/stable/config/integrating.html#MyObject._repr_mimebundle_
- Image
  - https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.Image
  - [ ] PNG (``_repr_png_()``)
  - [ ] **SVG** (``_repr_svg_()``)

ipywidgets
- https://ipywidgets.readthedocs.io/en/latest/examples/Using%20Interact.html
- https://ipywidgets.readthedocs.io/en/latest/examples/Output%20Widget.html

nbgrader
- auto-graded Jupyter notebooks
- https://github.com/jupyter/nbgrader
- is it possible to check design parameters with nbgrader?

.

- Is this in scope for this project?
- Are there any existing implementations with similar functionality?
- What sort of API does it need?
  - ``render(object_)``?

**Comment by looooo** [comment]
Yes we talked about displaying freecad objects in jupyger, but no work was done so far. We represent objects with pivs/coin Scenegraph library in OpenGL. It would be nice to reuse the viewobjects and draw the objects in webgl too. There is one library (osgjs) which implements the inventor scenegraph in JavaScript. I guess choosing this library for a webgl viewer of freecad objects would be the ideal way.
But there is also the possibility to get a vrml-represetation from freeecad and draw the objects with eg. https://github.com/maartenbreddels/ipyvolume. I think this is the simpler way. Maybe some monkey patching could be done to add the repr-functions to freecad-related...

**Comment by westurner** [comment]
- "osg.js widget"
https://groups.google.com/forum/m/#!topic/jupyter/73yz3w3vzxc
- "IPython notebook integration"
https://forum.freecadweb.org/viewtopic.php?t=14740

It looks like vispy does WebGL with Jupyter notebook. IDK if that's a
helpful reference for building an extension with osgjs.
https://github.com/vispy/vispy

ipyvolume looks great.

I got p5.js-svg working in Jupyter notebook awhile back.

A Jupyterlab extension is the best way to include JS in a notebook.
JupyterLab does not support inline script execution; but does have an
extension API.
http://jupyterlab.readthedocs.io/en/stable/developer/extension_dev.html
https://jupyterlab.readthedocs.io/en/stable/user/extensions.html

https://github.com/jupyterlab/jupyter-renderers
> Renderers and renderer extensions for JupyterLab

>
